lib/src/solver/solve_suggestions.dart

```diff
@@ -0,0 +1,214 @@
+
+import '../command_runner.dart';
+import '../entrypoint.dart';
+import '../flutter_releases.dart';
+import '../io.dart';
+import '../lock_file.dart';
+import '../package.dart';
+import '../package_name.dart';
+import '../pubspec_utils.dart';
+import '../solver.dart';
+import '../system_cache.dart';
+import 'incompatibility.dart';
+import 'incompatibility_cause.dart';
+
+/// Looks through the root-[incompability] of a solve-failure and tries to see if
+/// the conflict could resolved by any of the following suggestions:
+/// * An update of the current SDK.
+/// * Any single change to a package constraint.
+/// * Removing the bounds on all constraints, changing less than 5 dependencies.
+/// * Running `pub upgrade --major versions`.
+///
+/// Returns a formatted list of suggestions, or the empty String if no
+/// suggestions were found.
+Future<String?> suggestResolutionAlternatives(
+  Entrypoint entrypoint,
+  SolveType type,
+  Incompatibility incompatibility,
+  Iterable<String> unlock,
+  SystemCache cache,
+) async {
+  Future<_ResolutionSuggestion?> suggestSdkUpdate(SdkCause cause) async {
+    final sdkName = cause.sdk.identifier;
+    if (!(sdkName == 'dart' || (sdkName == 'flutter' && runningFromFlutter))) {
+      // Only make sdk upgrade suggestions for Flutter and Dart.
+      return null;
+    }
+
+    final constraint = cause.constraint;
+    if (constraint == null) return null;
+
+    /// Find the most relevant Flutter release fullfilling the constraint.
+    final bestRelease =
+        await inferBestFlutterRelease({cause.sdk.identifier: constraint});
+    if (bestRelease == null) return null;
+    final result = await tryResolve(type, cache, entrypoint.root,
+        lockFile: entrypoint.lockFile,
+        unlock: unlock,
+        sdkOverrides: {
+          'dart': bestRelease.dartVersion,
+          'flutter': bestRelease.flutterVersion
+        });
+    if (result != null) {
+      return _ResolutionSuggestion(
+        runningFromFlutter
+            ? '* Try using the Flutter SDK version: ${bestRelease.flutterVersion}. '
+            :
+            // Here we assume that any Dart version included in a Flutter
+            // release can also be found as a released Dart SDK.
+            '* Try using the Dart SDK version: ${bestRelease.dartVersion}. See https://dart.dev/get-dart.',
+      );
+    }
+    return null;
+  }
+
+  Future<_ResolutionSuggestion?> suggestSinglePackageUpdate(String name) async {
+    final originalConstraint = (entrypoint.root.dependencies[name] ??
+            entrypoint.root.devDependencies[name])
+        ?.constraint;
+    if (originalConstraint != null) {
+      final relaxedPubspec = stripVersionBounds(entrypoint.root.pubspec,
+          stripOnly: [name], stripLowerBound: true);
+
+      final result = await tryResolve(
+          type, cache, Package.inMemory(relaxedPubspec),
+          lockFile: entrypoint.lockFile, unlock: unlock);
+      if (result != null) {
```

Comment:
```suggestion
      if (result == null) {
        return;
      }
```
Then we get less indentation...

Reply:
Done